machines: Reimplement the design of the main VMs list

KKoukiou opened the pull request:
The expandable content of each row was moves to a seperate page.

Based on the design of #14196

Things that remain and need to be fixed in this PR
* [x] Adjust tests

---

_List of issues to address in follow-up PR(s) moved to #14604._

KKoukiou commented:
@garrett I think I am quite close to the suggested design, there are 1-2 more things that I would prefer to do in seperate PR because the tests are already going to need huge adjustment only by these changes.

![Screen Shot 2020-07-10 at 19 02 06](https://user-images.githubusercontent.com/14921356/87180005-fe68d480-c2df-11ea-8209-fc239ee03648.png)

KKoukiou commented on pkg/lib/cockpit-components-table.scss:
```diff
@@ -97,6 +97,17 @@
+
+        > .pf-m-link {
+            &:hover {
+                text-decoration: underline;
```
For underliner on hover we can use isInline property on the Button links instead of adding custom CSS.

garrett replied:
Oh! Yeah, that would be better.

Is there something that would handle the block expansion, or do we need custom CSS for that still?

KKoukiou replied:
We need custom CSS for that.

garrett commented:
Is it possible to let middle-click open the sub-pages in a separate tab? (We might need to make the link-styled buttons actually links. Or capture middle click events and do something a little more fancy?)

KKoukiou commented:
> Is it possible to let middle-click open the sub-pages in a separate tab? (We might need to make the link-styled buttons actually links. Or capture middle click events and do something a little more fancy?)

Done.

garrett commented:
Oh. Middle-clicking (and control+clicking, which should do the same thing) opens in the same page, not another tab (at least here). I'm only asking about having them show up in another page because I would imagine that people would want to see and work on multiple VMs at once.

KKoukiou commented:
@garrett this is ready from my side - do you want to take a look?
Definitely not inside todays release - but let's target next one?

martinpitt commented:
The code overall structure looks fine to me. Thanks! I'm just not a fan of the new design. It looks very untidy and unstructured now, at several resolutions.

On full width the console area is too crammed, and it feels like there's too many different fonts, sizes, and elements put on random places:

![image](https://user-images.githubusercontent.com/200109/91182594-a7ca2500-e6ea-11ea-9aa3-3460cc8ecffc.png)

Plus, the "Launch remote viewer" button should be PF4 styled.

When you switch to serial console, the window is very narrow:

![image](https://user-images.githubusercontent.com/200109/91182694-caf4d480-e6ea-11ea-8e1a-4842e1ef95b1.png)

On a narrower screen, the layout becomes more vertical, but it wastes a lot of space and many things scroll off:

![image](https://user-images.githubusercontent.com/200109/91182516-8bc68380-e6ea-11ea-9ab2-e0c790bede85.png)

martinpitt reviewed (changes requested):
See above; plus, there is some rebasing error in the second commit.

KKoukiou commented:
@garrett this needs your help, I can't make narrow screens work nicely.

garrett commented:
@martinpitt: Thanks for your review! I agree. This PR needs some layout adjustments. I'll look into it this week.

(Apologies; I had to deprioritze this PR to address the metrics designs and various other PRs that were closer to land.)

KKoukiou commented:
@garrett I was playing with it, we can also consider grey background:
![Screen Shot 2020-09-02 at 20 55 42](https://user-images.githubusercontent.com/14921356/92024763-e604bb80-ed5e-11ea-9d72-37b1f3239a9a.png)

garrett commented:
> I was playing with it, we can also consider grey background

Yeah, good idea. We're doing that across Cockpit already, and it does look nice here.

KKoukiou commented:
@garrett I just tested it interactively, the Disk and Network tables don't wrap, causing the mobile mode to be broken. Do you want me to take a look or should you address it?

garrett commented:
> @garrett I just tested it interactively, the Disk and Network tables don't wrap, causing the mobile mode to be broken. Do you want me to take a look or should you address it?

Yeah, I noticed that in the evening, before I committed this. I was already working too late and didn't want to just keep all of yesterday's changes locally, so I didn't address it yet. I'm going to today.

This is also part of the reason why I tried to split it up a bit. If it's one of the smaller commits, it'll be easier to figure out what's going on. If it's in the larger one (more likely), it won't be as nice to figure out.

KKoukiou commented:
> > @garrett I just tested it interactively, the Disk and Network tables don't wrap, causing the mobile mode to be broken. Do you want me to take a look or should you address it?
> 
> Yeah, I noticed that in the evening, before I committed this. I was already working too late and didn't want to just keep all of yesterday's changes locally, so I didn't address it yet. I'm going to today.
> 
> This is also part of the reason why I tried to split it up a bit. If it's one of the smaller commits, it'll be easier to figure out what's going on. If it's in the larger one (more likely), it won't be as nice to figure out.

It's the grid breakpoint - it's too small - I will adjust to gridBreakPoint: grid-xl

garrett commented:
Usage shouldn't be visible unless the machine is running.

This might impact the layout with console, but as it comes after usage in the document order, we could move the CSS grid span rule into something like `.usage + .console {}`.

KKoukiou commented:
> Usage shouldn't be visible unless the machine is running.
> 
> This might impact the layout with console, but as it comes after usage in the document order, we could move the CSS grid span rule into something like `.usage + .console {}`.

I feel it's weird that the whole layout changes when ones runs and stops the machines. Note that users often will stop/start the Vm being within that page.

garrett commented:
> I feel it's weird that the whole layout changes when ones runs and stops the machines. Note that users often will stop/start the Vm being within that page.

The layout doesn't change really, the usage would show up when it applies and the console would stretch to accommodate that.

I think it's weirder that there's (currently) a usage card with bogus stats and a very tall console card without anything in it.

garrett commented:
The Networking and Storage pages have since moved to a card-based layout, partially inspired by this Machines page. Since then, the buttons have been moved to the top-right of the cards. I think we might want to think about moving them to the top-right like on those pages too.

Additionally, the auto-start on the mockup had it at the header, at the top of the page, away from the details... similar to the service details pages.

Here's a modified mockup that takes these changes into account:

![vm-details](https://user-images.githubusercontent.com/10246/92613168-f1a23600-f2ba-11ea-9cdc-dcc7a8b4e49d.png)

garrett commented:
This is getting close to be able to merge.

Things still needed:

- [ ] hide usage when machine isn't running
- [ ] hide expand arrow on consle when machine isn't running
- [ ] move buttons back to the top-right of the cards
- [ ] change button priority (run should be primary, the others secondary)

Follow-up PR:
- move main VM list into a card

garrett commented:
Card dropshadows were missing; I removed the override to re-enable them in d5f3be8.

KKoukiou commented:
> Card dropshadows were missing; I removed the override to re-enable them in [d5f3be8](https://github.com/cockpit-project/cockpit/commit/d5f3be8a9a233f9ff7b76caf6e3e4bc25af11e0c).

Squashed in the main commit and resent.

garrett commented on pkg/machines/machines.scss:
```diff
@@ -155,11 +154,6 @@
     margin: 0.25rem 0 0.25rem 0.5rem;
 }
 
-.machines-network-list {
-    max-height: 500px;
-    padding-bottom: 15px;
-}
```
This was a remnant of the previous expand-o-list version of the VM page and it was getting in the way of the details page.

I wonder how much legacy CSS is still in the machines.scss? It's probably worth doing a cleanup as a follow-up PR.

garrett reviewed (dismissed):
:+1: LGTM design-wise! :tada: 

We'll handle a few additional issues in follow-up PRs.

martinpitt commented on test/verify/check-machines:
```diff
@@ -313,14 +313,8 @@ class TestMachines(MachineCase, StorageHelpers, NetworkHelpers):
-        b.wait_present("#chart-donut-1 .donut-title-big-pf")
-        self.assertLessEqual(get_usage("#chart-donut-1"), 100.0)
+        b.wait_in_text(".memory-usage-chart .pf-c-progress__status > .pf-c-progress__measure", "256 MiB")
+        b.wait_in_text(".vcpu-usage-chart .pf-c-progress__status > .pf-c-progress__measure", "1 vCPU")
```
Can you please validate that the usage goes above zero here?

martinpitt commented on pkg/machines/app.jsx:
```diff
@@ -128,6 +134,42 @@ class App extends React.Component {
+                                               onDismiss={() => this.onDismissErrorNotification(notification.index)}
+                                               title={notification.text}>{notification.detail}</Alert>
+                                    );
+                                }) : undefined}
```
I think this matches the `?` on line 148. It's hard to read, can you indent it properly and also start the line with `:` so that it lines up with the '?'

martinpitt commented on pkg/machines/app.jsx:
```diff
@@ -128,6 +134,42 @@ class App extends React.Component {
+                return null;
+
+            const connectionName = vm.connectionName;
+            const expandedContent = vm.isUi ? undefined : (
```
What does `isUi` actually do? I see it being set in `setVmCreateInProgress()` and `setVmInstallInProgress()`, so it sounds like some kind of "pending operation" flag?

**Update**: Ah, I get it now -- that's kind of a dummy placeholder until libvirt gets a real domain object for newly created VMs.

garrett replied:
> dummy placeholder until libvirt gets a real domain object for newly created VMs

This should probably be turned into a comment above `.isUI` usage in the code.

martinpitt commented on pkg/machines/app.jsx:
```diff
@@ -128,6 +134,42 @@ class App extends React.Component {
+                    interfaces={interfaces}
+                    networks={networks.filter(network => network && network.connectionName == connectionName)}
+                    nodeDevices={nodeDevices.filter(device => device && device.connectionName == connectionName)}
+                    key={`${vmId(vm.name)}`}
```
Does the 
```
$`...`
```
wrapper do some kind of magic here? that looks like an expensive no-op.

martinpitt commented on pkg/machines/components/vm/vmActions.jsx:
```diff
@@ -43,6 +54,61 @@ const VmActions = ({ vm, dispatch, storagePools, onStart, onInstall, onReboot, o
+    });
+    const onReboot = () => dispatch(rebootVm(vm)).catch(ex => {
+        onAddErrorNotification({
+            text: cockpit.format(_("VM $0 failed to Reboot"), vm.name),
```
lower-case "r"eboot, please. Same for the force variant below.

martinpitt commented on pkg/machines/components/vm/vmActions.jsx:
```diff
@@ -43,6 +54,61 @@ const VmActions = ({ vm, dispatch, storagePools, onStart, onInstall, onReboot, o
+    });
+    const onSendNMI = () => dispatch(sendNMI(vm)).catch(ex => {
+        onAddErrorNotification({
+            text: cockpit.format(_("VM $0 failed to force send NMI"), vm.name),
```
Let's drop the "force" here.

martinpitt commented on pkg/machines/components/vm/vmExpandedContent.jsx:
```diff
@@ -17,49 +17,170 @@
+            className: "consoles-card",
+            title: <>{_("Console")}<Button variant="link"
+                                           isDisabled={vm.state == "shut off"}
+                                           onClick={() => cockpit.location.go(["vm", "console"], Object.assign(cockpit.location.options, { name: vm.name, connection: vm.connectionName }))}
```
Very long line. Perhaps use the spread operator instead of Object.assign()

martinpitt commented on test/verify/check-machines:
```diff
@@ -102,8 +102,18 @@ class TestMachines(MachineCase, StorageHelpers, NetworkHelpers):
-    def toggleVmRow(self, vmName, connectionName='system'):
-        self.browser.click("tbody tr[data-row-id=vm-{0}-{1}] .pf-c-table__toggle button".format(vmName, connectionName)) # click on the row header
+        # HACK: https://launchpad.net/bugs/1802005
+        if m.image == "ubuntu-stable":
```
Is this really still the case? ubuntu-stable == ubuntu 20.10 prerelease now, that bug ought to be fixed. If not, we need to reopen it. But we dropped that hack from all other check-machines cases.

martinpitt reviewed (changes requested):
Really nice work, thanks! Great to remove c3, and we can even remove d3 as well (independently). I have a few minor nitpicks, but these should be fixable in 5 mins. Let's get a round of test results, though.

martinpitt commented on pkg/machines/components/vm/vmExpandedContent.jsx:
```diff
@@ -17,49 +17,167 @@
+        {
+            id: `${vmId(vm.name)}-consoles`,
+            className: "consoles-card",
+            title: <>{_("Console")}<Button variant="link"
```
This looks like only the icon would be clickable -- shouldn't the "Console" word be part of the link and get clickable as well?

garrett replied:
We should have text next to the icon that says "maximize" or "expand".

It's already mentioned in the follow-up PR @ https://github.com/cockpit-project/cockpit/issues/14604 — I'm waiting for this PR to land to address those issues.

martinpitt commented:
Several [testAddDisk failures](https://logs-https-cockpit.apps.ci.centos.org/logs/pull-14346-20200915-063635-5322aeee-fedora-33/log.html#263-2), looking.

martinpitt reviewed (dismissed):
Thanks @KKoukiou for all the fixes! LGTM now. I can reproduce the testAddDisk() flake locally, it failed 1 out of 2 cases. I retried it, if it does not flake too much I'm ok with fixing that as a follow-up, as this already blocks @garrett.

martinpitt commented:
I'm afraid it's pretty persistent. Also, the screenshot looks really broken:

![screenshot](https://logs-https-cockpit.apps.ci.centos.org/logs/pull-14346-20200915-131232-5322aeee-ubuntu-stable/TestMachines-testAddDisk-ubuntu-stable-127.0.0.2-2202-FAIL.png)

That suggests that the CSS needs some more work?

garrett commented:
@martinpitt: The screenshot is broken. I cannot see it; it's just a link here... and when I follow the link, I get "Error Fetching Resource". Can you re-upload the screenshot please?

martinpitt commented:
@garrett : It's just a link to the test failure's screenshot. I adjusted it in the previous comment to be a simple link instead of an embedded image.

garrett commented:
@martinpitt: Link works now. Thanks! That's... odd and not what I'm seeing at all. Weird. Admittedly, I was hurrying through the PR and didn't do cross-browser testing at the end of last week, so it might be some difference between Chrome and Firefox. I'll check (and fix) it now.

garrett commented:
I cannot recreate anything like in that screenshot.

What browser and version is that test using?!?

Chrome:
![Screenshot from 2020-09-16 11-42-36](https://user-images.githubusercontent.com/10246/93321431-6d980300-f812-11ea-9ae0-62dfbfe76f69.png)

WebKit;
![Screenshot from 2020-09-16 11-42-42](https://user-images.githubusercontent.com/10246/93321463-7983c500-f812-11ea-92ce-120f47047e10.png)
(It has some issue with the size of the Overview section being too narrow. This might be due to the way it's handling the overflow on the snapshots table, or perhaps it's because it's not sizing the console down.)

Granted, there are other issues, but I think it's mainly related to an overflow bug in the snapshots table (as seen in Chrome):
![Screenshot from 2020-09-16 11-42-50](https://user-images.githubusercontent.com/10246/93321544-94563980-f812-11ea-86c6-13bd5a0d7e56.png)

Note: I've tested non-running VMs and the layout is fine too.

Note 2: I can see snapshots locally, finally! This is one of the reasons I wanted to upgrade to F33.

---

The CSS issues I _do_ see: 
- [ ] Console header is not correct. (A regression compared to earlier.)
- [ ] Snapshots table is going out-of-bounds for some odd reason.
- [ ] WebKit is compressing the Overview. 
   - It might not be sizing down the console screen? 
   - Or it could be related to the snapshots out of bound?
   - (Either way, WebKit is doing something different here compared to Chrome and Firefox. It might be resolved with an OOB Snapshots fix.)

Non-CSS layout seen in the test:
- [ ] Remove button is flush to the right, due to the lack of an edit button.
   - (How does one get a remove without edit?) 
   - This could be fixed with a left-align for the buttons, or by having a disabled edit button (which could be set to be `visibility: hidden` — or not).

martinpitt commented:
@garrett: It's chromium-headless-81.0.4044.138-1.fc32.x86_64 . Thanks for validating again, I suggest we ignore that broken screenshot for now. The failing testDiskAdd still needs fixing, though, before we can land this.

martinpitt commented:
Tests still [show a crash](https://logs.cockpit-project.org/logs/pull-14346-20200916-120621-9359ff77-debian-testing/log.html#57-2)

martinpitt commented:
Trivial conflict now with #14629, sorry.

garrett commented:
Added a few fixes that I've been working on all day. The layout was glitchy and I was hitting browser bugs too.

It looks like WebKit doesn't support blocking inline styles (at least on Linux?)

After a bunch of testing, it seems this new breakpoint works better as well.

Additionally, I hope the expand link makes sense now. It's basically like a maximize button on a window, but with the word "expand" to the side. (Should we even call it "maximize"... or is that a bit odd? :thinking:)

garrett commented:
I've rebased this on master locally, fixed commit issues, and pushed to my fork. Commits here: https://github.com/cockpit-project/cockpit/compare/master...garrett:issue-14196

(I don't want to do a force-push in case I messed up the merge. Fixing the conflicts were _mostly_ straightforward, however.)

marusak commented:
> I've rebased this on master locally, fixed commit issues, and pushed to my fork

Please push here unless you think you made a mistake. We still can revert if there is some obvious problem. I am happy then to finish up tests and give this test. Thanks!

garrett commented:
OK; I've rebased on master again and force-pushed (with my merge conflict fixes from yesterday).

KKoukiou commented:
@marusak this has passed from martins review in the past, but there was an test failing which I fixed. Do you mind taking a look and maybe we can get it into tomorrows release.

marusak reviewed (approved):
Crazy cool! Love the new design :) 
Approving. Admittedly I didn't go through changes super thoroughly but I checked when Martin approved and what followed.
Retried tests to make sure there are no surprises. Lets merge this when we are confident enough :)

KKoukiou commented:
> Crazy cool! Love the new design :)
> Approving. Admittedly I didn't go through changes super thoroughly but I checked when Martin approved and what followed.
> Retried tests to make sure there are no surprises. Lets merge this when we are confident enough :)

yeah I would like to merge that after tomorrows release, just for being on the safe side.

garrett commented:
Somehow, the layout of the VM page broke a bit...
![Screenshot_2020-10-15 Virtual Machines - garrett Rain](https://user-images.githubusercontent.com/10246/96115778-c5945900-0ee7-11eb-91af-f408d5754d18.png)

The details is all wonky now. It's split, there's too much padding all over the place. I guess something changed in the past few days?

I'll look into this now and open up a PR (or an issue, if I can't quickly solve it) to address it.
